Update to bootstrap 5: adapt shortcode 'cardpane'

deining opened the pull request:
- Closes #1158, per https://github.com/google/docsy/pull/1376#issuecomment-1407464018
- Closes #1375

Preview: https://deploy-preview-1376--docsydocs.netlify.app/docs/adding-content/shortcodes/#card-panes

chalin commented:
Preview: https://deploy-preview-1376--docsydocs.netlify.app/docs/adding-content/shortcodes/#card-panes

chalin reviewed (commented):
Could you add an entry to the CHANGELOG, under breaking-changes so that users are aware of the class name change?

chalin commented:
This PR improves the rendering of card panes, but note that this change still doesn't bring the concerned page in visual agreement with what the paged looked like before.

Before (https://deploy-preview-1356--docsydocs.netlify.app/docs/adding-content/shortcodes/#card-panes):

> <img width="805" alt="image" src="https://user-images.githubusercontent.com/4140793/214599075-ab41e9f0-14d7-47fd-bdcb-c0cf69953891.png">

With this PR (https://deploy-preview-1376--docsydocs.netlify.app/docs/adding-content/shortcodes/#card-panes):

> <img width="840" alt="image" src="https://user-images.githubusercontent.com/4140793/214599375-dbe81bbb-323c-47f6-ac0d-0d3e124154d7.png">

deining commented:
> Could you add an entry to the CHANGELOG, under breaking-changes so that users are aware of the class name change?

Done.

deining commented:
> This PR improves the rendering of card panes, but note that this change still doesn't bring the concerned page in visual agreement with what the paged looked like before.

I was aware of this when submitting the PR.

From the [bootstrap migration guide](ttps://getbootstrap.com/docs/5.0/migration/#card)

```
BREAKING: Dropped .card-deck in favor of our grid.
Wrap your cards in column classes and add a parent .row-cols-* container
to recreate card decks (but with more control over responsive alignment).
```

Using of the bootstrap grid in the `cardpane` shortcode would mean a major rework of the shortcode. I fiddled around with adding a margin/padding to the individual cards, with no satisfactory results. Yes, the page looks different, but it doesn't look bad IMHO. I would leave this as is for now.

deining commented:
> This PR improves the rendering of card panes, but note that this change still doesn't bring the concerned page in visual agreement with what the paged looked like before.
 
On a second attempt, I now was able to bring the concerned page in visual agreement with what the page looked before.
The PR now also closes #1158.

chalin commented on CHANGELOG.md:
```diff
@@ -21,6 +21,10 @@ For a list of issues targeted for the next release, see the [23Q1][] milestone.
     - Dropped `$primary-light`.
     - Dropped `color-diff()`.
 
+- **[Adaptation of shortcode 'cardpane'][1376]. With the update to Bootstrap 5,
+  the class name used in the `cardpane` shortcode and the associated css file
+  had to be changed from `card-deck` to `card-group`.
```
```suggestion
- **[Adaptation of shortcode 'cardpane'][1376]: renamed CSS class `td-card-deck` to `td-card-group`.
```

deining replied:
Changed as proposed.

chalin reviewed (commented):
See inline comment.

Also pls rebase, then I'll do a final review.

deining commented:
> Also pls rebase,

Done.

> then I'll do a final review.

Thanks.

deining commented:
Following up on #906, I reworked the `card` shortcode by adding a second commit to this PR. This commit introduces the following changes:

-  shortcode `card`, `card-code`: markup of inner content (html/markdown) now depends on the syntax of the calling shortcode, not on extension of page file any more.
- deprecated shortcode `card-code`, you can now use shortcode `card` with named parameter `code=true` instead.
- Simplified and improved code of shortcodes.

chalin reviewed (commented):
Is `layouts/shortcodes/card-code.html` actually still used? It seems to me that this PR was getting rid of it, but the file is still present.

chalin commented on layouts/shortcodes/card-code.html:
```diff
@@ -1,20 +1,15 @@
+{{ $lang := default "" ($.Get "lang") -}}
+{{ $highlight := default "" ($.Get "highlight") -}}
+
+<div class="td-card card border me-4">
```
I thought that this shortcode was being dropped?

If not, then I'd suggest adding any custom styling through `cards-pace.scss`; avoid the `border` class here, and even consider moving the `me-4` into, say, `.td-card.card`.

deining replied:
> I thought that this shortcode was being dropped?

See my answer in this [commen](https://github.com/google/docsy/pull/1376#issuecomment-1429333570)

deining replied:
> If not, then I'd suggest adding any custom styling through `cards-pace.scss`; avoid the `border` class here, and even consider moving the `me-4` into, say, `.td-card.card`.

I agree. I will open a new issue that addresses this issue shortly.

chalin commented on layouts/shortcodes/card.html:
```diff
@@ -1,49 +1,44 @@
+{{ $lang := default "" (.Get "lang") -}}
+{{ $highlight := default "" (.Get "highlight") -}}
+
+<div class="td-card card border me-4">
```
As I mentioned before, I'd suggest adding custom styling through `cards-pace.scss`; avoid the `border` class here, and even consider moving the `me-4` into, say, `.td-card.card`.

deining replied:
> As I mentioned before, I'd suggest adding custom styling through `cards-pace.scss`; avoid the `border` class here, and even consider moving the `me-4` into, say, `.td-card.card`.

I agree. I will open a new issue that addresses this issue shortly.

chalin commented on layouts/shortcodes/card.html:
```diff
@@ -1,49 +1,44 @@
-      {{ else }}
-        {{ $.Get "header" | htmlUnescape | safeHTML }}
-      {{ end }}
+{{/* Check parameter type */ -}}
+{{ with .Get "code" -}}
+  {{ if ne ( printf "%T" . ) "bool" -}}
+    {{ errorf "shortcode card: parameter 'code' must be either true or false" -}}
+  {{ end -}}
+{{ end -}}
```
Is the type check really necessary? Why not interpret the value simply as truthy / falsy?

deining replied:
This type check is here for a good reason: the correct syntax is `<card code=true>`. If you write `<card code="true">`, an error ist thrown. What happens if you omit the check? See shortcode `readfile` which has no type check (yet!?):

https://github.com/google/docsy/blob/3eb2cd0cfb26521d94cc8503e0092cc4a5c40cc2/layouts/shortcodes/readfile.html#L31

Users are expected to write `<readfile code="true">`, if they write `<readfile code=true>`, nothing happens. This will be surprising for users and I doubt that all users immediately see what they did wrong (giving parameter value `true` instead of `"true"`).

Does that make clear why I added the type check? Does this answer your question and resolve the issue?

**Additional note**: IMHO shortcode `readfile` does things wrong here: it treats parameter `code` as string and tests for string equality while the parameter `code` has boolean character and should be treated like that in the shortcode.

chalin replied:
Yes, that makes sense. In that case, the error message could be clearer, stating that a boolean value is expected, but another type was given. E.g.,

> {{ errorf "shortcode card: boolean value expected for parameter 'code', but got type %s" ( printf "%T" . ) -}}

You can even extract the value of `( printf "%T" . )` earlier and reuse it in both places.

deining replied:
Improved error message as requested.

chalin commented on layouts/shortcodes/card.html:
```diff
@@ -1,49 +1,44 @@
+
+<div class="td-card card border me-4">
+{{ with $.Get "header" -}}
+    <div class="card-header{{- cond $code " bg-white" "" -}}">
```
Here's another place where we could avoid hard-coding custom styling, but for now I'm ok if we leave this as is, and address it later.

deining replied:
> Here's another place where we could avoid hard-coding custom styling,

I agree.

> but for now I'm ok if we leave this as is, and address it later.

I will open a new issue that addresses this issue shortly.

chalin reviewed (commented):
Other than the question about `card-code` and border, LGTM. Thanks!

deining commented:
> Is `layouts/shortcodes/card-code.html` actually still used? 

No, in the userguide, we don't use it any more.

> It seems to me that this PR was getting rid of it, but the file is still present.

CHANGELOG:

```
- shortcode `card-code` is now deprecated, use shortcode `card` with named
    parameter `code=true` instead.
```

I wrote `deprecated`, which means the file is still present and works, but it is not documented any more and users are discouraged to further use this shortcode.

> but the file is still present.

Yes, this is intentional: if I remove the shortcode, this will break all existing pages that make use of this shortcode and we will force users to adapt their pages (by replacing shortcode `card-code` with `card`).

Does that clarify things? Would you like to see the file deleted?

deining commented:
> Other than the question about `card-code` and border, LGTM. Thanks!

I rebased the PR to get it ready for merging. Let me know if you want me to delete the `card-code` shortcode. Styling issues will be addressed in a separate issue, to be opened shortly. Thanks for your thorough review, I appreciate your comments.

chalin commented on layouts/shortcodes/card-code.html:
```diff
@@ -1,20 +1,15 @@
-    {{ $highlight = $.Get "highlight" }}
-  {{ end }}
-  {{- with $.Get "header" -}}
+{{ $lang := default "" ($.Get "lang") -}}
```
Consider adding at least a comment to the effect that this is deprecated.

I had missed the deprecation note int the changelog, so others might miss it too. I propose that you add a warning stating that it is deprecated and will be removed in a future release. This way projects using the shortcode will get notified when they build their site. WDYT?

deining replied:
> Consider adding at least a comment to the effect that this is deprecated.

Great idea, I will do that shortly.

deining replied:
Done.

chalin reviewed (commented):
Thanks for your clarifications. See inline comments.

chalin reviewed (approved):
I've rebased and resolved the minor conflicts. Thanks for the progressive improvements to this PR. 

LGTM ✨
